pyteal/ast/router.py

```diff
@@ -0,0 +1,583 @@
+            )
+        method_signature = method_call.method_signature(overriding_name)
+        if call_configs.is_never():
+            raise TealInputError(
+                f"registered method {method_signature} is never executed"
+            )
+        if method_signature in self.added_method_sig:
+            raise TealInputError(f"re-registering method {method_signature} detected")
+        self.added_method_sig.add(method_signature)
+
+        wrapped = Router._wrap_handler(True, method_call)
+
+        create_has_clear_state = call_configs._has_clear_state(CallConfig.CREATE)
+        create_others = call_configs._approval_program_oc(CallConfig.CREATE)
+        call_has_clear_state = call_configs._has_clear_state(CallConfig.CALL)
+        call_others = call_configs._approval_program_oc(CallConfig.CALL)
+
+        if create_has_clear_state:
+            self.categorized_clear_state_ast.method_calls_create.append(
+                CondNode(
+                    And(
+                        Txn.application_id() == Int(0),
+                        Txn.application_args[0] == MethodSignature(method_signature),
+                    ),
+                    wrapped,
+                )
+            )
+        if call_has_clear_state:
+            self.categorized_clear_state_ast.method_calls.append(
+                CondNode(
+                    Txn.application_args[0] == MethodSignature(method_signature),
+                    wrapped,
+                )
+            )
+
+        if create_others:
+            self.categorized_approval_ast.method_calls_create.append(
+                CondNode(
+                    And(
+                        Txn.application_id() == Int(0),
+                        Txn.application_args[0] == MethodSignature(method_signature),
+                        Or(*[Txn.on_completion() == oc for oc in create_others]),
+                    ),
+                    wrapped,
+                )
+            )
+        if call_others:
+            self.categorized_approval_ast.method_calls.append(
+                CondNode(
+                    And(
+                        Txn.application_args[0] == MethodSignature(method_signature),
+                        Or(*[Txn.on_completion() == oc for oc in call_others]),
+                    ),
+                    wrapped,
+                )
+            )
+
+    def contract_construct(self) -> dict[str, Any]:
+        """A helper function in constructing contract JSON object.
+
+        It takes out the method signatures from approval program `ProgramNode`'s,
+        and constructs an `Contract` object.
+
+        Returns:
+            contract: a dictified `Contract` object constructed from
+                approval program's method signatures and `self.name`.
+        """
+        method_collections = [
+            sdk_abi.Method.from_signature(sig) for sig in self.added_method_sig
+        ]
+        return sdk_abi.Contract(self.name, method_collections).dictify()
```

Comment:
Why not return an `sdk_abi.Contract` object? I think that would be a little more useful

Reply:
👍

Reply:
👍

Reply:
👍